src/lib.rs

```diff
@@ -1258,30 +1121,109 @@ pub enum Descriptor {
+        let ptr = ptr.try_borrow().unwrap_or_else(|_| unreachable());
+        let ret = f(&*ptr);
+        match ret {
+            Ok(()) => ERRNO_SUCCESS,
+            Err(err) => err,
+        }
+    }
+
+    fn with_mut(f: impl FnOnce(&mut State) -> Result<(), Errno>) -> Errno {
+        let ptr = State::ptr();
+        let mut ptr = ptr.try_borrow_mut().unwrap_or_else(|_| unreachable());
+        let ret = f(&mut *ptr);
+        match ret {
+            Ok(()) => ERRNO_SUCCESS,
+            Err(err) => err,
+        }
+    }
+
+    fn ptr() -> &'static RefCell<State> {
+        assert!(size_of::<State>() <= PAGE_SIZE);
         unsafe {
-            let mut fds = replace_fds(null_mut());
-            if fds.is_null() {
-                let grew = core::arch::wasm32::memory_grow(0, 1);
-                if grew == usize::MAX {
-                    unreachable();
+            match get_global_ptr() {
+                0 => {
+                    let grew = core::arch::wasm32::memory_grow(0, 1);
+                    if grew == usize::MAX {
+                        unreachable();
+                    }
+                    let ret = (grew * PAGE_SIZE) as *mut RefCell<State>;
+                    set_global_ptr(ret as usize);
+                    ret.write(RefCell::new(State {
+                        buffer_ptr: Cell::new(null_mut()),
+                        buffer_len: Cell::new(0),
+                        ndescriptors: 0,
+                        descriptors: MaybeUninit::uninit(),
+                        path_buf: MaybeUninit::uninit(),
+                    }));
+                    &*ret
                 }
-                fds = (grew * PAGE_SIZE) as *mut u8;
+                1 => unreachable(),
+                other => &*(other as *const RefCell<State>),
             }
-            // We allocated a page; abort if that's not enough.
-            if fd as usize * size_of::<Descriptor>() >= PAGE_SIZE {
-                unreachable()
+        }
+    }
+
+    fn push_desc(&mut self, desc: Descriptor) -> Result<Fd, Errno> {
+        unsafe {
+            let descriptors = self.descriptors.as_mut_ptr();
+            if self.ndescriptors >= (*descriptors).len() {
+                return Err(ERRNO_INVAL);
             }
-            let result = &mut *fds.cast::<Descriptor>().add(fd as usize);
-            let fds = replace_fds(fds);
-            assert!(fds.is_null());
-            result
+            ptr::addr_of_mut!((*descriptors)[self.ndescriptors]).write(desc);
+            self.ndescriptors += 1;
+            Ok(Fd::try_from(self.ndescriptors - 1).unwrap())
+        }
+    }
+
+    fn get(&self, fd: Fd) -> Result<&Descriptor, Errno> {
+        let index = usize::try_from(fd).unwrap();
+        if index < self.ndescriptors {
+            unsafe { (*self.descriptors.as_ptr()).get(index).ok_or(ERRNO_BADF) }
+        } else {
+            Err(ERRNO_BADF)
+        }
+    }
+
+    fn get_file(&self, fd: Fd) -> Result<&File, Errno> {
```

Comment:
Different functions are expected to return different errors.
 - Functions which take an offset: fd_advise, fd_seek, and fd_tell, should return `ERRNO_SPIPE` on a Log.
 - path_* functions should return `ERRNO_NOTDIR` on a Log.
 
Would it make sense to add a `get_seekable_file` and `get_dir` for these, which work similar to `get_file` here, but with this different errno codes?

Reply:
Sounds good to me, I ended up changing the error code for `Log` with `fd_filestat_get`, `fd_filestat_set_size`, and `fd_filestat_set_times`. Those previously were either `NOTDIR` or `SPIPE` and now they all return `INVAL` for `Log` descriptions. Additionally `fd_{pread,read}` return `INVAL` instead of `BADF` now.